File: 프로그래머스/난이도별/Level2.게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n, m = len(maps), len(maps[0])
    
    queue = deque()
    queue.append((0,0))
    
    visited = [[False for _ in range(m)] for _ in range(n)]
    dx = [1,-1,0,0]
    dy = [0,0,-1,1]

    while queue:
        x, y = queue.popleft()
        visited[x][y] = True
        
        if x == n-1 and y == m-1:
            return maps[n-1][m-1]
        
        for i in range(4):
            nx, ny = dx[i] + x, dy[i] + y
            if nx >= 0 and ny >= 0 and nx < n and ny < m:
                if maps[nx][ny] == 1 and not visited[nx][ny]:
                    maps[nx][ny] = maps[x][y] + 1
                    visited[nx][ny] = True
                    queue.append((nx,ny))
    
    return -1

File: 프로그래머스/난이도별/Level2.게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
  const n = maps.length;
  const m = maps[0].length;

  const dx = [-1, 1, 0, 0];
  const dy = [0, 0, -1, 1];

  const queue = [[0, 0, 1]];

  while (queue.length) {
    const item = queue.shift();
    const x = item[0];
    const y = item[1];
    const cnt = item[2];

    if (x === n - 1 && y === m - 1) {
      return cnt;
    }

    for (let i = 0; i < 4; i++) {
      const nx = dx[i] + x;
      const ny = dy[i] + y;
      if (nx > -1 && ny > -1 && nx < n && ny < m && maps[nx][ny]) {
        maps[nx][ny] = 0;
        queue.push([nx, ny, cnt + 1]);
      }
    }
  }
  return -1;
}
